socs/agents/acu/drivers.py

```diff
@@ -80,123 +95,140 @@ def get_track_points_text(tpl, timestamp_offset=None, with_group_flag=False,
+          number of points in the track (i.e. in each vector) that are
+          not included in the loopable portion.  This permits the
+          track to have a ramp-up, or partial initial scan segment,
+          prior to entering the repetable template.
 
-    all_azflags[-1] = 0
-
-    return conctimes, concaz, concel, concva, concve, all_azflags, \
-        all_elflags
-
-
-def from_file(filename):
     """
-    Produces properly formatted lists of times, azimuth and elevation
-    locations, azimuth and elevation velocities, and azimuth and elevation
-    motion flags for a finitely long scan from a numpy file. Numpy file
-    must be formatted as an array of arrays in the order [times, azimuths,
-    elevations, azimuth velocities, elevation velocities]
-
-    Parameters:
-        filename (str): Full path to the numpy file containing scan
-            parameter array
+    if fmt is None:
+        fmt = 'pickle'
+        if filename.endswith('npy'):
+            fmt = 'numpy'
+
+    if fmt == 'numpy':
+        info = np.load(filename)
+        if len(info) not in [5, 7]:
+            raise ValueError(f'Unexpected field count ({len(info)}) in {filename}')
+        times, az, el, vaz, vel = info[:5]
+        if len(info) == 5:
+            az_flags = np.zeros(len(times), int)
+            el_flags = az_flags
+        elif len(info) == 7:
+            az_flags = info[5].astype('int')
+            el_flags = info[6].astype('int')
+
+        output = FromFileScan()
+        output.loop_time = 0.
+        output.free_form = False
+        output.step_time = np.diff(times).min()
+        output.points = [TrackPoint(*a) for a in zip(
+            times, az, el, vaz, vel, az_flags, el_flags)]
+        output.az_range = (az.min(), az.max())
+        output.el_range = (el.min(), el.max())
+
+    elif fmt == 'pickle':
+        data = pickle.load(open(filename, 'rb'))
+        output = FromFileScan()
+        output.loop_time = 0.
+        output.free_form = data.get('free_form', False)
+
+        keys = ['timestamp', 'az', 'el', 'az_vel', 'el_vel',
+                'az_flag', 'el_flag', 'group_flag']
+        vects = {k: data.get(k) for k in keys}
+        n = len(vects['az'])
+
+        dtv = np.gradient(vects['timestamp'])
+        if vects['az_vel'] is None:
+            vects['az_vel'] = np.gradient(vects['az']) / dtv
+        if vects['el_vel'] is None:
+            vects['el_vel'] = np.gradient(vects['el']) / dtv
+        if vects['az_flag'] is None:
+            vects['az_flag'] = np.zeros(n, int)
+        if vects['el_flag'] is None:
+            vects['el_flag'] = np.zeros(n, int)
+        if vects['group_flag'] is None:
+            vects['group_flag'] = np.zeros(n, int)
+
+        output.preamble_count = data.get('preamble_count', 0)
+        if data.get('loopable'):
+            # Measure repeat time.
+            output.loop_time = np.dot([-1, 1], vects['timestamp'][[output.preamble_count, -1]])
```

Comment:
This syntax seems to assume `vects['timestamp']` is a `np.array`. Based on the docs I would have thought a plain `list` would work here.